### PR TITLE
tests: add integration tests for transactional boundaries

### PR DESCRIPTION
**Pull request by nancyxu123**
tests: Add an integration test for change stream transaction boundaries
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+import org.junit.runners.JUnit4;
+
+/** End-to-end test of Cloud Spanner Change Streams Transaction Boundaries. */
+@RunWith(JUnit4.class)
+public class SpannerChangeStreamTransactionBoundariesIT {
+
+  private static final int MAX_TABLE_NAME_LENGTH = 128;
+  private static final int MAX_CHANGE_STREAM_NAME_LENGTH = 30;
+  private static final String TABLE_NAME_PREFIX = "Singers";
+  public static final String SPANNER_HOST = "https://staging-wrenchworks.sandbox.googleapis.com";
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java:80]: Could we add a `TODO` to use the production host instead? In fact, can't we use the production one already?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+    // Insert  live mutations after the pipeline already started running.
+
+    // 4. Commit a transaction to insert Singer 7 and update Singer 6 in the table.
+    mutations.add(insertRecordMutation(7, "FirstName7", "LastName7"));
+    mutations.add(updateRecordMutation(6, "FirstName5", "LastName5"));
+    Timestamp t4 = databaseClient.write(mutations);
+    System.out.println("The fourth transaction committed with timestamp: " + t4.toString());
+    mutations.clear();
+
+    Thread.sleep(5_000);
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java:231]: Why do we need to sleep here?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+
+    // 5. Commit a transaction to update Singer 4 and Singer 5 in the table.
+    mutations.add(updateRecordMutation(4, "FirstName9", "LastName9"));
+    mutations.add(updateRecordMutation(5, "FirstName9", "LastName9"));
+    Timestamp t5 = databaseClient.write(mutations);
+    System.out.println("The fifth transaction committed with timestamp: " + t5.toString());
+    mutations.clear();
+
+    Thread.sleep(5_000);
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java:240]: Why do we need to sleep here?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java [hunk 4]
```diff
@@ -0,0 +1,528 @@
...
+    // 6. Commit a transaction to delete Singers 3, 4, 5.
+    mutations.add(deleteRecordMutation(3));
+    mutations.add(deleteRecordMutation(4));
+    mutations.add(deleteRecordMutation(5));
+    Timestamp t6 = databaseClient.write(mutations);
+    System.out.println("The sixth transaction committed with timestamp: " + t6.toString());
+
+    // Assert that the returned PCollection contains all six transactions (in string representation)
+    // and that each transaction contains, in order, the list of mutations added.
+    PAssert.that(tokens).containsInAnyOrder(
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java:251]: I would imagine that there is a race condition here, where we might not get all records. Is this true, could this be flaky? If so, I would wait for some condition to be true before performing the assertion (maybe the number of records == some number?)

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java (hunk 1 above)
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java:80]: Sure, let me give that a try.

I'm assuming we already have a change stream demo database in production? What is the project / instance / database name for production? I don't think I have used it before actually.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java (hunk 2 above)
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java:231]: I added a sleep there because I thought it could enhance testing for live writes, but you're right, it's actually not needed. Removing in next commit.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java (hunk 3 above)
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java:240]: See above.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java (hunk 4 above)
Comment (nancyxu123) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java:251]: From what I understand about PAssert, it seems that the PAssert is part of the pipeline graph and validates the items in the PCollection globally (i.e. across all windows / triggers)? I am under that assumption, because so far there has been no flakiness, and it seems that PAssert has other functions that specifically validate for windows: https://beam.apache.org/releases/javadoc/2.0.0/org/apache/beam/sdk/testing/PAssert.IterableAssert.html#containsInAnyOrder-java.lang.Iterable-. 

Also, there is documentation stating that PAssert must be called before pipeline.run(): https://beam.apache.org/releases/javadoc/2.0.0/org/apache/beam/sdk/testing/PAssert.html

One thing I am thinking about however, is the flakiness that might be caused by the 4th, 5th, or 6th transactions committing after the pipeline stops running (i.e. if those transaction timestamps are after end timestamps). I haven't seen any failures yet, but maybe if there was some slowness in testing threads, this could happen. It doesn't really seem possible however, to wait for (number of records == some number), since I don't really know how we would extract the number of records?

I was thinking perhaps we could just not set an end timestamp. Then, after the sixth transaction commits, we could cancel pipeline execution after some delay (i.e. 30 seconds)?

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java (hunk 1 above)
Comment (thiagotnunes) [reply sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java:81]: We actually don't have one as we were not testing in production (other than manual testing). 
This is the default configuration for beam in terms of Spanner (https://github.com/thiagotnunes/beam/blob/plan-a/sdks/java/io/google-cloud-platform/build.gradle#L165-L175). We can either allowlist this one or we can use https://pantheon.corp.google.com/spanner/instances/changestream/databases/changestream/details/tables?project=span-cloud-testing, which is the one we are using for manual testing at the moment

**Review by thiagotnunes — COMMENTED** [review]
Could you please change your PR title to:

`tests: add integration tests for transactional boundaries`

(for context we usually follow https://www.conventionalcommits.org/en/v1.0.0/#summary, although it is not required here)

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+
+    List<Mutation> mutations = new ArrayList<>();
+    // Insert historical mutations into the data table to be returned by the Change Stream
+    // Connector.
+
+    // 1. Commit a transaction to insert Singer 1 and Singer 2 into the table.
+    mutations.add(insertRecordMutation(1, "FirstName1", "LastName2"));
+    mutations.add(insertRecordMutation(2, "FirstName2", "LastName2"));
+    Timestamp t1 = databaseClient.write(mutations);
+    System.out.println("The first transaction committed with timestamp: " + t1.toString());
```
Comment (thiagotnunes) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/spanner/SpannerChangeStreamTransactionBoundariesIT.java:181]: @nancyxu123 Could we use a logger instead of `System.out`?

```
private static final Logger LOG = LoggerFactory.getLogger(SpannerChangeStreamTransactionBoundariesIT.class);

LOG.debug(...)
```

**Review by thiagotnunes — APPROVED** [review]
